shpc/main/registry/remote.py

```diff
@@ -0,0 +1,182 @@
+            return response.text
+
+    def override_exists(self, tag):
+        """
+        Check if an override exists.
+        """
+        logger.exit(
+            "We don't have support for remote overrides yet! Please open an issue!"
+        )
+
+
+class VersionControl(Provider):
+    def __init__(self, *args, **kwargs):
+        self.tag = kwargs.get("tag")
+
+        # Cache of remote container metadata
+        self._cache = {}
+
+        # E.g., subdirectory with registry files
+        self.subdir = kwargs.get("subdir")
+        super().__init__(*args, **kwargs)
+        self._url = self.source
+
+    @classmethod
+    def matches(cls, source):
+        return cls.provider_name in source and source.startswith("http")
+
+    def exists(self, name):
+        """
+        Determine if a module exists in the registry.
+        """
+        dirname = self.source
+        if self.subdir:
+            dirname = os.path.join(dirname, self.subdir)
+        return os.path.exists(os.path.join(dirname, name))
+
+    def clone(self, tmpdir=None):
+        """
+        Clone the known source URL to a temporary directory
+        """
+        tmpdir = tmpdir or shpc.utils.get_tmpdir()
+
+        cmd = ["git", "clone", "--depth", "1"]
+        if self.tag:
+            cmd += ["-b", self.tag]
+        cmd += [self._url, tmpdir]
+        self.source = tmpdir
+        try:
+            sp.run(cmd, check=True)
+        except sp.CalledProcessError as e:
+            raise ValueError("Failed to clone repository {}:\n{}", self.source, e)
+        return tmpdir
+
+    def iter_modules(self):
+        """
+        yield module names
+        """
+        dirname = self.source
+        if self.subdir:
+            dirname = os.path.join(dirname, self.subdir)
+
+        # Find modules based on container.yaml
+        for filename in shpc.utils.recursive_find(dirname, "container.yaml"):
+            module = os.path.dirname(filename).replace(dirname, "").strip(os.sep)
+            if not module:
+                continue
+            yield dirname, module
+
+    def find(self, name):
+        """
+        Find a particular entry in a registry
+        """
+        self._update_cache()
+        if name in self._cache:
+            return RemoteResult(name, self._cache[name])
+
+    def _update_cache(self, force=False):
+        """
+        Update local cache from a registry.
+        """
+        if self._cache and not force:
+            return
+        if not self.source.startswith("http"):
+            self.source = "https://%s" % self.source
+
+        # Check for API
+        org, repo = self.source.split("/")[3:]
+        gh_pages = "https://%s.%s.io/%s/library.json" % (org, self.provider_name, repo)
+        response = requests.get(gh_pages)
+        if response.status_code != 200:
+            sys.exit(
+                "Remote %s is not deploying a Registry API. Open a GitHub issue to ask for help."
+                % self.source
+            )
```

Comment:
```suggestion
        url = self.source
        if not url.startswith("http"):
           url = "https://%s" % url
        if url.endswith(".git"):
            url = source[:-4]

        # Check for API
        url_parts = url.split("/")[3:]
        gh_pages = "https://%s.%s.io/%s/library.json" % (url_parts[0], self.provider_name, "/".join(url_parts[1:]))
        response = requests.get(gh_pages)
        if response.status_code != 200:
            sys.exit(
                "Remote %s is not deploying a Registry API (%s). Open a GitHub issue to ask for help."
                % (self.source, gh_pages)
            )
```

Reply:
Can you walk me through this change? Why wouldn't we want to change the source (for good) with a proper https url?

Reply:
About this particular change, I don't understand why the source would not start with `https://` at this stage, since the constructor checks that and raises an error.
That's why I had suggested to remove it in https://github.com/singularityhub/singularity-hpc/pull/579/files#diff-183e41e47a9fe06125f88dd47d33eb5bc399a6458ad3bc0e8d5c98e422179239L147-L148

Reply:
I did that because a person may just want to specify it as `github.com/<org>/<name>` . I don't see that the constructor checks and throws an error, it seems to just set it: https://github.com/singularityhub/singularity-hpc/blob/132d04ad4e41abaafd96c762961f48735e5de682/shpc/main/registry/remote.py#L86.

Reply:
The check happens in the super-class https://github.com/singularityhub/singularity-hpc/blob/f948e4ecffc43dc595bdc7e6b62298399f6a02ef/shpc/main/registry/provider.py#L35-L40

Reply:
But there is an edge case, if they are in a go src directory and there is `github.com` it will pass (and then later fail).

Reply:
This is a more of a "quick bail out before we do anything" check.

Reply:
But not to get distracted with this check, which could be argued is redundant - can you explain the changes in the above? What problem is it fixing? I think we are close and (things look generally ok) and could merge/release soon.

Reply:
The patch is to allow support for "groups" in gitlab, cf your message (https://github.com/singularityhub/singularity-hpc/pull/576#issuecomment-1221642688):
> and in the meantime if you want to share the update to match "org" and "repo"

Reply:
with groups, the url is https://gitlab.com/group/sub-group/repo.git

Reply:
Ah gotcha, thank you for clarifying! Let me see if I can organize the functions a bit - brb!

Reply:
okay check out https://github.com/singularityhub/singularity-hpc/pull/576/commits/1f29489067e6aa639b214a829bcd57aabf3f62ca. !